### PR TITLE
fix: build musa

### PR DESCRIPTION
**Pull request by yeahdongcn**
This PR enables building for MUSA (S4000).

### CODE REVIEW

**Comment by yeahdongcn** [comment]
@thxCode Could you please run the workflow to check if the build succeeds? Thanks!

File: .github/workflows/ci.yml [hunk 1]
```diff
@@ -475,6 +475,89 @@ jobs:
           path: ${{ github.workspace }}/out/*.zip
           name: llama-box-linux-${{ matrix.arch }}-cann-${{ matrix.version }}
 
+  linux-musa:
+    strategy:
+      fail-fast: false
+      matrix:
+        arch: [ amd64 ]
+        version: [ '1.5.1-mp_22' ]
```
Comment (thxCode) [inline .github/workflows/ci.yml:483]: can we use a '{major}.{minor}' version here? and then map below, as other jobs did?

just like `version: [ '1.5' ]`

File: README.md [hunk 2]
```diff
@@ -12,20 +12,21 @@ than [llama-server](https://github.com/ggerganov/llama.cpp/blob/master/examples/
...
+| Intel oneAPI 2024.1        | `linux/amd64` `windows/amd64` | See Linux/Windows distributions with oneAPI 2024.1 from https://www.intel.com/content/www/us/en/developer/articles/system-requirements/intel-oneapi-base-toolkit-system-requirements.html.                              |
+| Intel oneAPI 2024.2        | `linux/amd64` `windows/amd64` | See Linux/Windows distributions with oneAPI 2024.2 from https://www.intel.com/content/www/us/en/developer/articles/system-requirements/intel-oneapi-base-toolkit-system-requirements.html.                              |
+| Ascend CANN 8.0            | `linux/amd64` `linux/arm64`   | See https://www.hiascend.com/document/detail/zh/CANNCommunityEdition/80RC3alpha001/quickstart/quickstart/quickstart_18_0002.html.                                                                                       |
+| Moore Threads MUSA 1.5.1   | `linux/amd64`                 | `MTT S4000`.                                                                                                                                                                                                            |
```
Comment (thxCode) [inline README.md:29]: can we provide a document or link here? so users can get more information from here.

**Comment by yeahdongcn** [comment]
The CMake version has been updated in the latest commit to align with our base image, Ubuntu 20.04.

File: .github/workflows/ci.yml (hunk 1 above)
Comment (yeahdongcn) [reply .github/workflows/ci.yml]: Done.

File: README.md (hunk 2 above)
Comment (yeahdongcn) [reply README.md:29]: This document is still a work in progress. I can update it later in a new PR, or we can remove it for now. Which option do you prefer?

Comment (thxCode) [reply README.md:29]: I think you can update it later. thanks.

**Comment by yeahdongcn** [comment]
The project hierarchy differs slightly from `Ollama`. `Ollama` injects its `ext_server` as a sub-target of `llama.cpp`, allowing `ext_server` to inherit all the compilation settings from it.

**Comment by thxCode** [comment]
yes, the box is not in-tree structure, so we can observe the compilation detail. I think we almost achieved the goal. can you add this and have a try?

```diff
diff --git a/llama-box/CMakeLists.txt b/llama-box/CMakeLists.txt
index 80dcd90..172e673 100644
--- a/llama-box/CMakeLists.txt
+++ b/llama-box/CMakeLists.txt
@@ -77,6 +77,16 @@ if (GGML_CANN)
                 ${CANN_INSTALL_DIR}/runtime/lib64/stub)
     endif ()
 endif ()
+if (GGML_MUSA)
+    set(CMAKE_C_COMPILER clang)
+    set(CMAKE_C_EXTENSIONS OFF)
+    set(CMAKE_CXX_COMPILER clang++)
+    set(CMAKE_CXX_EXTENSIONS OFF)
+
+    set(GGML_CUDA ON)
+
+    list(APPEND GGML_CDEF_PUBLIC GGML_USE_MUSA)
+endif()
 add_executable(${TARGET} main.cpp param.hpp ratelimiter.hpp utils.hpp)
 add_dependencies(${TARGET} patch)
 target_link_libraries(${TARGET} PRIVATE version common llava ${CMAKE_THREAD_LIBS_INIT})

```

based on your work, I tried it on my laptop, is it green?

image

**Comment by yeahdongcn** [comment]
Yes, I also tried switching the default `C/CXX` compiler to `Clang` locally, and the compilation finished without errors. However, it seems odd that `GGML_MUSA` is used directly. Is this acceptable to you? 😂

**Comment by thxCode** [comment]
> Yes, I also tried switching the default `C/CXX` compiler to `Clang` locally, and the compilation finished without errors. However, it seems odd that `GGML_MUSA` is used directly. Is this acceptable to you? 😂

LGTM at present.

**Comment by yeahdongcn** [comment]
> ```diff
> +    set(CMAKE_C_COMPILER clang)
> +    set(CMAKE_C_EXTENSIONS OFF)
> +    set(CMAKE_CXX_COMPILER clang++)
> +    set(CMAKE_CXX_EXTENSIONS OFF)
> ```

Cool! Please see the latest commit, thanks.

**Comment by yeahdongcn** [comment]
The MUSA build is now successful.

**Review by thxCode — APPROVED** [review]
LGTM

**Pull request merged**